### ISSUE TITLE
external_ip: add option to choose between IPv4 and IPv6 (if available)

### ISSUE THREAD

**Issue by hgigas** [greshake/i3status-rust#1761]
```
[[block]]
block = "external_ip"
format = "$ip"
```

shows: `2a02:9130:fd....`

**Comment by kagaroth** [comment]
This is not a bug. In dual stack you use IPv6 first then you fallback to IPv4 and because the site / server used to test and show your public IP uses both IPv4 and IPv6 and because you also have an IPv6 set on your active connection the you use that v6 IP for communicating with it and that is what is shown in the output of that block.

**Comment by hgigas** [comment]
I'm sorry, it isn't a bug.

What's my ip web page shows me both IPv6 and IPv4. How to display only IPv4 in `external_ip` block?

**Comment by kagaroth** [comment]
Put this is `/etc/hosts`:
```
104.26.8.44 ipapi.co
104.26.9.44 ipapi.co
172.67.69.226 ipapi.co
```
This should force the use of IPv4 when talking to ipapi.co as this disable recursive DNS resolution for ipapi .co IP. That way your machine will know that ipapi.co is the 3 IPv4 listed above in `/etc/hosts`.

Before:
```
{
    "ip": "2a02:2f0f:b2ff:ffff::5679:5aad",
    "network": "2a02:2f0f:b000::/38",
    "version": "IPv6",
...
}
```
After:
```
{
    "ip": "86.121.90.173",
    "network": "86.121.88.0/22",
    "version": "IPv4",
...
}
```

**Comment by GladOSkar** [comment]
@kagaroth's solution works but is a bit messy (means no other application can obtain ipv6 from ipapi.com anymore). We should probably provide an option for this.

**Comment by hgigas** [comment]
That would be wonderful, thank you very much.

Thank you a lot @kagaroth for the fast reply

**Comment by cfsmp3** [comment]
> That would be wonderful, thank you very much.
> 
Could you try this, see if it does what you want and then we'll see how to make it properly?

Edit `main.rs` and add on line 56 (after the user_agent one), this:

```
 .local_addr("0.0.0.0:0".parse().unwrap())
```
so it looks like this:

```
   reqwest::Client::builder()
        .user_agent(APP_USER_AGENT)
         .local_addr("0.0.0.0:0".parse().unwrap())
        .timeout(REQWEST_TIMEOUT)
        .build()
        .unwrap()
``` 
 
This forces ipv4 in all calls, but for testing at least it should be good enough...

**Comment by hgigas** [comment]
The code:

```rust
pub static REQWEST_CLIENT: Lazy<reqwest::Client> = Lazy::new(|| {
    const APP_USER_AGENT: &str = concat!(env!("CARGO_PKG_NAME"), "/", env!("CARGO_PKG_VERSION"),);
    const REQWEST_TIMEOUT: Duration = Duration::from_secs(10);
    reqwest::Client::builder()
        .user_agent(APP_USER_AGENT)
        .local_addr("0.0.0.0:0".parse().unwrap())
        .timeout(REQWEST_TIMEOUT)
        .build()
        .unwrap()
});
```

raises error:

```rust
error[E0599]: no method named `local_addr` found for struct `ClientBuilder` in the current scope
  --> src/main.rs:58:10
   |
58 |         .local_addr("0.0.0.0:0".parse().unwrap())
   |          ^^^^^^^^^^ help: there is a method with a similar name: `local_address`

For more information about this error, try `rustc --explain E0599`.
error: could not compile `i3status-rs` due to previous error
```

**Comment by cfsmp3** [comment]
> The code:
> 
> ```rust
> pub static REQWEST_CLIENT: Lazy<reqwest::Client> = Lazy::new(|| {
>     const APP_USER_AGENT: &str = concat!(env!("CARGO_PKG_NAME"), "/", env!("CARGO_PKG_VERSION"),);
>     const REQWEST_TIMEOUT: Duration = Duration::from_secs(10);
>     reqwest::Client::builder()
>         .user_agent(APP_USER_AGENT)
>         .local_addr("0.0.0.0:0".parse().unwrap())
>         .timeout(REQWEST_TIMEOUT)
>         .build()
>         .unwrap()
> });
> ```
> 
> raises error:
> 
> ```rust
> error[E0599]: no method named `local_addr` found for struct `ClientBuilder` in the current scope
>   --> src/main.rs:58:10
>    |
> 58 |         .local_addr("0.0.0.0:0".parse().unwrap())
>    |          ^^^^^^^^^^ help: there is a method with a similar name: `local_address`
> 
> For more information about this error, try `rustc --explain E0599`.
> error: could not compile `i3status-rs` due to previous error
> ```

Well, do what it says and use local_address instead of local_addr :-)

**Comment by hgigas** [comment]
I did it and then another error fixed adding `::<IpAddr>` to `parse` and then only display short format in all blocks... 

> @kagaroth's solution works but is a bit messy (means no other application can obtain ipv6 from ipapi.com anymore). We should probably provide an option for this.

IMO is a very good idea.

**Comment by kagaroth** [comment]
Just a thought, the `external_ip` block only needs to return the IPv4 because NAT is used only in version 4. For version 6 users can use the `net` block as it already has `$ip6` as placeholder. IPv6 was made with the idea of not using NAT anymore.

**Comment by GladOSkar** [comment]
@kagaroth there do exist IPv6 setups with NAT. I wouldn't make that kind of assumption

**Comment by kagaroth** [comment]
Yes there are, I don't know who and why they would do that (NAT66) nor how widely deployed those setups are. I'm not going to assume why they exist, not referring to NAT64 or NAT46. I was talking from the point of view of a home user that uses a standard setup with a "SOHO" router that gets a /56 or /64 prefix from service provider and allocates IPs in LAN from that "pool". 

I am sorry for the rant.

**Comment by cfsmp3** [comment]
I think it's safe to expect our user base to have non standard network
setups both at home and at work 😁

I strongly oppose to dumbing things down.

**Comment by GladOSkar** [comment]
> I did it and then another error fixed adding ::<IpAddr> to parse and then only display short format in all blocks and external_ip hidden...

That's because there was a `AddrParseError` because an `IpAddr` is not supposed to have a port. This is the correct diff, and works:
```diff
@@ -55,6 +55,7 @@ pub static REQWEST_CLIENT: Lazy<reqwest::Client> = Lazy::new(|| {
     const REQWEST_TIMEOUT: Duration = Duration::from_secs(10);
     reqwest::Client::builder()
         .user_agent(APP_USER_AGENT)
+        .local_address("0.0.0.0".parse::<std::net::IpAddr>().unwrap())
         .timeout(REQWEST_TIMEOUT)
         .build()
```

But i still don't like this. It enforces IPv4 for all blocks using HTTP and also is initialized before the config is loaded (i think). Maybe a solution would be to just always initialize 2 instances of `reqwest` (one for v4, one for v6) (at least if both versions are available on any interfaces) and then choose which one to use based on the block config? Seems a bit messy as well. But i'm not that good at rust, maybe someone has a better idea?

**Comment by MaxVerevkin** [comment]
> initialized before the config is loaded (i think)

It is initialized on first use.

> Maybe a solution would be to just always initialize 2 instances of reqwest (one for v4, one for v6) 

Yes, we can have two lazy instances of reqwest, one with default IP config and one with IPv4 forced. The good thing is that the second instance is not created if it is not used.

BTW, there's `Ipv4Addr::UNSPECIFIED`, so no parsing is needed.

**Comment by MaxVerevkin** [comment]
Implemented in #1801
